File: internal/game/mapgen/maze_test.go
package mapgen

import (
	"testing"

	gamedata "github.com/gravestench/dark-magic/internal/game/data/catalog"
	model "github.com/gravestench/dark-magic/internal/game/data/model"
)

func mazeFixture() gamedata.Snapshot {
	presets := make([]model.LevelPreset, 0, 15)
	byDef := make(map[int]model.LevelPreset, 15)
	for mask := 1; mask <= 15; mask++ {
		record := model.LevelPreset{Def: 52 + mask, SizeX: 24, SizeY: 24, Files: 1, File1: "Act1/Caves/room.ds1", Dt1Mask: 1}
		presets = append(presets, record)
		byDef[record.Def] = record
	}
	return gamedata.Snapshot{
		LevelsByID:       map[int]model.LevelData{9: {Id: 9, Act: 0, DrlgType: 1, LevelType: 3}},
		LevelMazeByLevel: map[int]model.LevelMazeData{9: {Level: 9, Rooms: 12, RoomsN: 14, RoomsH: 16, SizeX: 24, SizeY: 24, Merge: 500}},
		LevelPresets:     presets, LevelPresetByDef: byDef,
		LevelTypes: []model.LevelType{{}, {}, {}, {File1: "Act1/Caves/floor.dt1"}},
	}
}

func TestMazeGeneratorCreatesConnectedNonOverlappingCompatibleRooms(t *testing.T) {
	zone, err := NewMazeGenerator(mazeFixture()).Generate(Request{Version: ContractVersion, Seed: 42, Act: 1, LevelID: 9})
	if err != nil {
		t.Fatal(err)
	}
	rooms, links, stamps := zone.Rooms(), zone.Links(), zone.Stamps()
	if len(rooms) != 12 || len(links) < len(rooms)-1 || len(stamps) != len(rooms) {
		t.Fatalf("rooms=%d links=%d stamps=%d", len(rooms), len(links), len(stamps))
	}
	seen := map[[2]int]bool{}
	for index, room := range rooms {
		key := [2]int{room.X, room.Y}
		if seen[key] {
			t.Fatalf("overlapping room at %v", key)
		}
		seen[key] = true
		if room.Width != 24 || room.Height != 24 || stamps[index].PresetDef < 53 || stamps[index].PresetDef > 67 {
			t.Fatalf("room/stamp %d = %#v / %#v", index, room, stamps[index])
		}
	}
	visited := map[uint32]bool{1: true}
	for changed := true; changed; {
		changed = false
		for _, link := range links {
			if visited[link.From] && !visited[link.To] {
				visited[link.To], changed = true, true
			}
			if visited[link.To] && !visited[link.From] {
				visited[link.From], changed = true, true
			}
		}
	}
	if len(visited) != len(rooms) {
		t.Fatalf("only %d/%d rooms connected", len(visited), len(rooms))
	}
}

func TestMazeGeneratorIsChecksumStableAndDifficultyAware(t *testing.T) {
	generator := NewMazeGenerator(mazeFixture())
	request := Request{Version: ContractVersion, Seed: 99, Act: 1, LevelID: 9, Difficulty: Nightmare}
	left, err := generator.Generate(request)
	if err != nil {
		t.Fatal(err)
	}
	right, err := generator.Generate(request)
	if err != nil {
		t.Fatal(err)
	}
	leftSum, _ := left.Checksum()
	rightSum, _ := right.Checksum()
	if leftSum != rightSum || len(left.Rooms()) != 14 {
		t.Fatalf("checksums %s/%s rooms=%d", leftSum, rightSum, len(left.Rooms()))
	}
}

func TestConnectionMaskMapsDirectlyToAuthoredCaveDefinition(t *testing.T) {
	masks := map[mazeCell]uint8{}
	center := mazeCell{}
	for _, neighbor := range []mazeCell{{-1, 0}, {1, 0}, {0, 1}, {0, -1}} {
		applyConnection(masks, center, neighbor)
	}
	if masks[center] != 15 || 52+int(masks[center]) != 67 {
		t.Fatalf("center mask=%#x preset=%d", masks[center], 52+int(masks[center]))
	}
}

File: internal/content/d2legacy/lua/d2legacy/mapgen/maze_test.lua
local test = require("d2legacy.tests/v1")

return test.suite({
    profile = "policy",
    tier = "fast",
    tests = {
        owns_act_one_cave_maze_policy = {
            {
                run = function()
                    local prest = {}
                    for definition = 53, 67 do
                        prest[#prest + 1] = {
                            Def = tostring(definition),
                            SizeX = "24",
                            SizeY = "24",
                            Files = "1",
                            File1 = "Act1/Caves/room.ds1",
                            Dt1Mask = "1",
                        }
                    end
                    for definition = 83, 90 do
                        prest[#prest + 1] = {
                            Def = tostring(definition),
                            SizeX = "24",
                            SizeY = "24",
                            Files = "1",
                            File1 = "Act1/Caves/special.ds1",
                            Dt1Mask = "1",
                        }
                    end
                    local rows = {
                        ["data/global/excel/Levels.txt"] = {
                            { Id = "9", Act = "0", DrlgType = "1", LevelType = "3" },
                        },
                        ["data/global/excel/LvlMaze.txt"] = {
                            {
                                Level = "9",
                                Rooms = "12",
                                ["Rooms(N)"] = "14",
                                ["Rooms(H)"] = "16",
                                SizeX = "24",
                                SizeY = "24",
                                Merge = "500",
                            },
                        },
                        ["data/global/excel/LvlTypes.txt"] = {
                            {},
                            {},
                            {},
                            { ["File 1"] = "Act1/Caves/floor.dt1" },
                        },
                        ["data/global/excel/LvlPrest.txt"] = prest,
                    }
                    package.loaded["engine.records/v1"] = {
                        load = function(path)
                            return rows[path] or {}
                        end,
                    }
                    package.loaded["d2legacy.mapgen.maze"] = nil
                    local maze = require("d2legacy.mapgen.maze")
                    local zone = maze.generate(9, 42, 0)
                    assert(zone.kind == "maze" and #zone.rooms == 12 and #zone.stamps == 12 and #zone.links >= 11)
                    local previous, next_level = 0, 0
                    local occupied = {}
                    for _, stamp in ipairs(zone.stamps) do
                        if stamp.role == "previous-level" then
                            previous = previous + 1
                        end
                        if stamp.role == "next-level" then
                            next_level = next_level + 1
                        end
                        local key = stamp.x .. ":" .. stamp.y
                        assert(not occupied[key])
                        occupied[key] = true
                    end
                    assert(previous == 1 and next_level == 1)
                    assert(zone.checksum == maze.generate(9, 42, 0).checksum)
                    assert(#maze.generate(9, 42, 1).rooms == 14)
                end,
            },
        },
    },
})
